app/src/main/cpp/skyline/kernel/scheduler.cpp

```diff
@@ -0,0 +1,385 @@
+
+    Scheduler::Scheduler(const DeviceState &state) : state(state) {}
+
+    void Scheduler::SignalHandler(int signal, siginfo *info, ucontext *ctx, void **tls) {
+        if (*tls) {
+            const auto &state{*reinterpret_cast<nce::ThreadContext *>(*tls)->state};
+            state.scheduler->Rotate(false);
+            YieldPending = false;
+            state.scheduler->WaitSchedule();
+        } else {
+            YieldPending = true;
+        }
+    }
+
+    Scheduler::CoreContext &Scheduler::GetOptimalCoreForThread(const std::shared_ptr<type::KThread> &thread) {
+        auto *currentCore{&cores.at(thread->coreId)};
+
+        if (!currentCore->queue.empty() && thread->affinityMask.count() != 1) {
+            // Select core where the current thread will be scheduled the earliest based off average timeslice durations for resident threads
+            // There's a preference for the current core as migration isn't free
+            size_t minTimeslice{};
+            CoreContext *optimalCore{};
+            for (auto &candidateCore : cores) {
+                if (thread->affinityMask.test(candidateCore.id)) {
+                    u64 timeslice{};
+
+                    if (!candidateCore.queue.empty()) {
+                        std::lock_guard coreLock(candidateCore.mutex);
+
+                        auto threadIterator{candidateCore.queue.cbegin()};
+                        if (threadIterator != candidateCore.queue.cend()) {
+                            const auto &runningThread{*threadIterator};
+                            timeslice += [&]() {
+                                if (runningThread->averageTimeslice)
+                                    return std::min(runningThread->averageTimeslice - (util::GetTimeTicks() - runningThread->timesliceStart), 1UL);
+                                else if (runningThread->timesliceStart)
+                                    return util::GetTimeTicks() - runningThread->timesliceStart;
+                                else
+                                    return 1UL;
+                            }();
+
+                            while (++threadIterator != candidateCore.queue.cend()) {
+                                const auto &residentThread{*threadIterator};
+                                if (residentThread->priority <= thread->priority)
+                                    timeslice += residentThread->averageTimeslice ? residentThread->averageTimeslice : 1UL;
+                            }
+                        }
+                    }
+
+                    if (!optimalCore || timeslice < minTimeslice || (timeslice == minTimeslice && &candidateCore == currentCore)) {
+                        optimalCore = &candidateCore;
+                        minTimeslice = timeslice;
+                    }
+                }
+            }
+
+            if (optimalCore != currentCore) {
+                /*
```

Comment:
?????

Reply:
I forgot to remove this prior to committing